parsing/tools/src/main/java/com/apache/calcite/parsing/tools/FindDialectParsingErrors.java

```diff
@@ -0,0 +1,424 @@
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.OutputStreamWriter;
+import java.lang.reflect.Method;
+import java.nio.charset.StandardCharsets;
+import java.text.MessageFormat;
+import java.text.ParseException;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+/**
+ * Processes a CSV file containing queries and will output a JSON file
+ * containing the results of failing queries.
+ */
+public class FindDialectParsingErrors {
+
+  private final String inputPath;
+  private final String outputPath;
+  private final Dialect dialect;
+  private final boolean groupByErrors;
+  private final int numSampleQueries;
+  private final List<MessageFormat> errorFormats;
+
+  /**
+   *  Creates a new instance of {@code FindDialectParsingErrors}. Also populates
+   *  the errorFormats list with a {@code MessageFormat} object for each custom
+   *  defined error message in CalciteResource.
+   *
+   * @param inputPath Path to the input CSV file containing queries
+   * @param outputPath Path to the output JSON file containing results of
+   *                   failing queries
+   * @param dialect Specifies which dialectic parser to use for processing
+   *                queries
+   * @param groupByErrors Specifies if the output format should group failing
+   *                      queries by error message type
+   * @param numSampleQueries The max number of sample queries to show for an
+   *                         error type when grouping queries by error message
+   *                         type
+   */
+  public FindDialectParsingErrors(String inputPath, String outputPath,
+      Dialect dialect, boolean groupByErrors, int numSampleQueries) {
+    this.inputPath = inputPath;
+    this.outputPath = outputPath;
+    this.dialect = dialect;
+    this.groupByErrors = groupByErrors;
+    this.numSampleQueries = numSampleQueries;
+    errorFormats = new ArrayList<>();
+    Method[] methods = CalciteResource.class.getMethods();
+    for (Method m : methods) {
+      errorFormats.add(
+          new MessageFormat(
+              m.getAnnotationsByType(
+              Resources.BaseMessage.class)[0].value(), Locale.ROOT)
+      );
+    }
+  }
+
+  /**
+   * Runs one of the processing methods and will output a JSON file containing
+   * the results of failing queries. This will run findErrorGroups if
+   * groupByErrors is true. Otherwise, findFullErrors will run.
+   *
+   * @throws IOException If it fails to read the input file
+   */
+  public void run() throws IOException {
+    CSVReader reader = new CSVReader(
+        new InputStreamReader(new FileInputStream(inputPath),
+        StandardCharsets.UTF_8), ',', '"', 1);
+    List<String[]> rows = reader.readAll();
```

Comment:
Will there potentially be a memory issue if the reader puts all of the queries in memory at the same time instead of streaming the queries?

Reply:
There might be if the file is extremely large. However, I'm not sure if I can use the paralellStream method without reading in the entire file at once. I think the only other option would be to read queries in one at a time and process them sequentially but that might slow it down.

Reply:
I think if I processed it sequentially it might still use the same amount of memory because it needs to store all the results since it sorts it at the end for the group by method

Reply:
Thanks for pointing this out, Lisa! My vote would be to keep FileIO as simple as possible for now until we encounter a problem here. As is, I've probably already complicated things enough by requesting that Michael process the queries in parallel :)

The targeted use case is to run through O(thousands) of queries at a time, which *shouldn't* require too much memory. However, if we find this assumption not to be true in practice or that target use case changes, then I think it might be worth revisiting this.

Reply:
Maybe consider leaving a comment here about it to help pinpoint where it might cause issues in the future?

Reply:
Yeah that would be a good idea, I'll leave a comment here